source/validate.cpp

```diff
@@ -261,6 +273,173 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
   return SPV_SUCCESS;
 }
 
+// TODO(umar): Validate header
+// TODO(umar): The Id bound should be validated also. But you can only do that
+// after you've seen all the instructions in the module.
+// TODO(umar): The binary parser validates the magic word, and the length of the
+// header, but nothing else.
+spv_result_t setHeader(void* user_data, spv_endianness_t endian, uint32_t magic,
+                       uint32_t version, uint32_t generator, uint32_t id_bound,
+                       uint32_t reserved) {
+  (void)user_data;
+  (void)endian;
+  (void)magic;
+  (void)version;
+  (void)generator;
+  (void)id_bound;
+  (void)reserved;
+  return SPV_SUCCESS;
+}
+
+// TODO(umar): Move this class to another file
+class ValidationState_t {
+ public:
+  ValidationState_t(spv_diagnostic* diag)
+      : diagnostic_(diag), instruction_counter_(0) {}
+
+  spv_result_t definedIds(uint32_t id) {
+    if (defined_ids_.find(id) == std::end(defined_ids_)) {
+      defined_ids_.insert(id);
+    } else {
+      return diag(SPV_ERROR_INVALID_ID)
+             << "ID cannot be assigned multiple times";
+    }
+    return SPV_SUCCESS;
+  }
+
+  spv_result_t forwardDeclareId(uint32_t id) {
+    unresolved_forward_ids_.insert(id);
+    return SPV_SUCCESS;
+  }
+
+  spv_result_t removeIfForwardDeclared(uint32_t id) {
+    unresolved_forward_ids_.erase(id);
+    return SPV_SUCCESS;
+  }
+
+  size_t unresolvedForwardIdCount() const {
+    return unresolved_forward_ids_.size();
+  }
+
+  //
+  bool isDefinedId(uint32_t id) const {
+    return defined_ids_.find(id) != std::end(defined_ids_);
+  }
+
+  // Increments the instruction count. Used for diagnostic
+  int incrementInstructionCount() { return instruction_counter_++; }
+
+  libspirv::DiagnosticStream diag(spv_result_t error_code) const {
+    return libspirv::DiagnosticStream({0, 0, (uint64_t)(instruction_counter_)},
+                                      diagnostic_, error_code);
+  }
+
+ private:
+  spv_diagnostic* diagnostic_;
+  // Tracks the number of instructions evaluated by the validator
+  int instruction_counter_;
+
+  // All IDs which have been defined
+  unordered_set<uint32_t> defined_ids_;
+
+  // IDs which have been forward declared but have not been defined
+  unordered_set<uint32_t> unresolved_forward_ids_;
+};
+
+// Performs SSA validation on the IDs of an instruction. The
+// can_have_forward_declared_ids  functor should return true if the
+// instruction operand's ID can be forward referenced.
+spv_result_t SsaPass(ValidationState_t& _,
```

Comment:
I just realized this strategy won't work.
The problem is it never computes dominators in the control flow.
The trouble is that your SsaPass algorithm will think any result-id for a previous instruction is valid to use.  But it doesn't distinguish between previous instructions that dominate this use vs. those that don't.

Here's an example where it would fail.  It's roughly structured as an if-then-else block, but where a value created in the "then" block is used in the "else" block.


```
   %conditionBlock = Label 
                                  SelectionMerge %merge None
                                  BranchConditional %condition %thenBlock %elseBlock

  %thenBlock = Label
                         %CreatedInThen = OpIAdd %i32 %one %two
                         Branch %merge

 %elseBlock  = Label
                        %BAD = OpCopyObject %CreatedInThen         ;;;;;; This is wrong. 
                                                                ;;;;;;;  Since %CreatedInThen definition does not dominate this basic block.
                        Branch %merge

   %merge = Label
  
```

Even more severe, a value defined in one function body would be seen as usable in the next function body.